### ISSUE TITLE
sbt 0.13.10-SNAPSHOT fails to compile akka/akka

### ISSUE THREAD

**Issue by Duhemm** [sbt/sbt#2263]
While testing #2261 on a large codebase (akka), sbt kept crashing. I first thought it was because of my changes, but it turns out that sbt cannot compile akka since #2085 and #2214 have been merged.
# Steps
1. Checkout sbt/sbt@33478132c5d0f4e109d85522ba52fd77bf470a9d
2. build and publish sbt locally.
3. Clone akka/akka and checkout akka/akka@18a24c0e5036ad97281732e87f33aea214caf8e3 (note: It looks like compiling akka requires java 8)
4. Compile akka using the freshly built sbt (change/remove `project/build.properties`).
# Problem

Compilation fails with the following stack trace:

```
java.lang.IllegalStateException: Failed to parse class java.util.AbstractMap$SimpleImmutableEntry: this may mean your classfiles are corrupted. Please clean and try again.
    at sbt.ClassToAPI$.fieldToDef(ClassToAPI.scala:162)
    at sbt.ClassToAPI$$anonfun$8.apply(ClassToAPI.scala:78)
    at sbt.ClassToAPI$$anonfun$8.apply(ClassToAPI.scala:78)
    at sbt.ClassToAPI$$anonfun$mergeMap$1.apply(ClassToAPI.scala:227)
    at sbt.ClassToAPI$$anonfun$mergeMap$1.apply(ClassToAPI.scala:227)
    at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:251)
    at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:251)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
    at scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:34)
    at scala.collection.TraversableLike$class.flatMap(TraversableLike.scala:251)
    at scala.collection.AbstractTraversable.flatMap(Traversable.scala:105)
    at sbt.ClassToAPI$.merge(ClassToAPI.scala:233)
    at sbt.ClassToAPI$.mergeMap(ClassToAPI.scala:227)
    at sbt.ClassToAPI$.structure(ClassToAPI.scala:78)
    at sbt.ClassToAPI$.x$2$lzycompute$1(ClassToAPI.scala:66)
    at sbt.ClassToAPI$.x$2$1(ClassToAPI.scala:66)
    at sbt.ClassToAPI$.instance$lzycompute$1(ClassToAPI.scala:66)
    at sbt.ClassToAPI$.sbt$ClassToAPI$$instance$1(ClassToAPI.scala:66)
    at sbt.ClassToAPI$$anonfun$5.apply(ClassToAPI.scala:67)
    at sbt.ClassToAPI$$anonfun$5.apply(ClassToAPI.scala:67)
    at xsbti.SafeLazy$Impl._t$lzycompute(SafeLazy.scala:18)
    at xsbti.SafeLazy$Impl._t(SafeLazy.scala:16)
    at xsbti.SafeLazy$Impl.get(SafeLazy.scala:22)
    at sbt.ClassToAPI$$anonfun$process$1.apply(ClassToAPI.scala:22)
    at sbt.ClassToAPI$$anonfun$process$1.apply(ClassToAPI.scala:22)
    at scala.collection.immutable.List.foreach(List.scala:318)
    at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:32)
    at scala.collection.mutable.ListBuffer.foreach(ListBuffer.scala:45)
    at sbt.ClassToAPI$.process(ClassToAPI.scala:22)
    at sbt.compiler.javac.AnalyzingJavaCompiler.sbt$compiler$javac$AnalyzingJavaCompiler$$readAPI$1(AnalyzingJavaCompiler.scala:76)
    at sbt.compiler.javac.AnalyzingJavaCompiler$$anonfun$compile$2$$anonfun$apply$mcV$sp$2$$anonfun$apply$5.apply(AnalyzingJavaCompiler.scala:84)
    at sbt.compiler.javac.AnalyzingJavaCompiler$$anonfun$compile$2$$anonfun$apply$mcV$sp$2$$anonfun$apply$5.apply(AnalyzingJavaCompiler.scala:84)
    at sbt.classfile.Analyze$$anonfun$apply$9.apply(Analyze.scala:44)
    at sbt.classfile.Analyze$$anonfun$apply$9.apply(Analyze.scala:43)
    at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:772)
    at scala.collection.mutable.HashMap$$anonfun$foreach$1.apply(HashMap.scala:98)
    at scala.collection.mutable.HashMap$$anonfun$foreach$1.apply(HashMap.scala:98)
    at scala.collection.mutable.HashTable$class.foreachEntry(HashTable.scala:226)
    at scala.collection.mutable.HashMap.foreachEntry(HashMap.scala:39)
    at scala.collection.mutable.HashMap.foreach(HashMap.scala:98)
    at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:771)
    at sbt.classfile.Analyze$.apply(Analyze.scala:43)
    at sbt.compiler.javac.AnalyzingJavaCompiler$$anonfun$compile$2$$anonfun$apply$mcV$sp$2.apply(AnalyzingJavaCompiler.scala:84)
    at sbt.compiler.javac.AnalyzingJavaCompiler$$anonfun$compile$2$$anonfun$apply$mcV$sp$2.apply(AnalyzingJavaCompiler.scala:82)
    at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:772)
    at scala.collection.immutable.List.foreach(List.scala:318)
    at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:771)
    at sbt.compiler.javac.AnalyzingJavaCompiler$$anonfun$compile$2.apply$mcV$sp(AnalyzingJavaCompiler.scala:82)
    at sbt.compiler.javac.AnalyzingJavaCompiler$$anonfun$compile$2.apply(AnalyzingJavaCompiler.scala:82)
    at sbt.compiler.javac.AnalyzingJavaCompiler$$anonfun$compile$2.apply(AnalyzingJavaCompiler.scala:82)
    at sbt.compiler.javac.AnalyzingJavaCompiler.timed(AnalyzingJavaCompiler.scala:93)
    at sbt.compiler.javac.AnalyzingJavaCompiler.compile(AnalyzingJavaCompiler.scala:81)
    at sbt.compiler.MixedAnalyzingCompiler$$anonfun$compileJava$1$1.apply$mcV$sp(MixedAnalyzingCompiler.scala:60)
    at sbt.compiler.MixedAnalyzingCompiler$$anonfun$compileJava$1$1.apply(MixedAnalyzingCompiler.scala:60)
    at sbt.compiler.MixedAnalyzingCompiler$$anonfun$compileJava$1$1.apply(MixedAnalyzingCompiler.scala:60)
    at sbt.compiler.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:74)
    at sbt.compiler.MixedAnalyzingCompiler.compileJava$1(MixedAnalyzingCompiler.scala:59)
    at sbt.compiler.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:64)
    at sbt.compiler.IC$$anonfun$compileInternal$1.apply(IncrementalCompiler.scala:160)
    at sbt.compiler.IC$$anonfun$compileInternal$1.apply(IncrementalCompiler.scala:160)
    at sbt.inc.IncrementalCompile$$anonfun$doCompile$1.apply(Compile.scala:66)
    at sbt.inc.IncrementalCompile$$anonfun$doCompile$1.apply(Compile.scala:64)
    at sbt.inc.IncrementalCommon.cycle(IncrementalCommon.scala:31)
    at sbt.inc.Incremental$$anonfun$1.apply(Incremental.scala:62)
    at sbt.inc.Incremental$$anonfun$1.apply(Incremental.scala:61)
    at sbt.inc.Incremental$.manageClassfiles(Incremental.scala:89)
    at sbt.inc.Incremental$.compile(Incremental.scala:61)
    at sbt.inc.IncrementalCompile$.apply(Compile.scala:54)
    at sbt.compiler.IC$.compileInternal(IncrementalCompiler.scala:160)
    at sbt.compiler.IC$.incrementalCompile(IncrementalCompiler.scala:138)
    at sbt.Compiler$.compile(Compiler.scala:152)
    at sbt.Compiler$.compile(Compiler.scala:138)
    at sbt.Defaults$.sbt$Defaults$$compileIncrementalTaskImpl(Defaults.scala:853)
    at sbt.Defaults$$anonfun$compileIncrementalTask$1.apply(Defaults.scala:844)
    at sbt.Defaults$$anonfun$compileIncrementalTask$1.apply(Defaults.scala:842)
    at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
    at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
    at sbt.std.Transform$$anon$4.work(System.scala:63)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
    at sbt.Execute.work(Execute.scala:235)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
    at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.RuntimeException: Could not create directory /Library/Java/JavaVirtualMachines/jdk1.8.0_60.jdk/Contents/Home/jre/lib/rt.jar/java/util
    at scala.sys.package$.error(package.scala:27)
    at sbt.IO$.createDirectory(IO.scala:177)
    at sbt.OpenFile$class.open(Using.scala:42)
    at sbt.Using$$anon$2.open(Using.scala:64)
    at sbt.Using$$anon$2.open(Using.scala:64)
    at sbt.Using.apply(Using.scala:23)
    at sbt.classfile.Parser$.apply(Parser.scala:18)
    at sbt.ClassToAPI$.sbt$ClassToAPI$$classFileForClass(ClassToAPI.scala:94)
    at sbt.ClassToAPI$.cf$lzycompute$1(ClassToAPI.scala:76)
    at sbt.ClassToAPI$.sbt$ClassToAPI$$cf$1(ClassToAPI.scala:76)
    at sbt.ClassToAPI$$anonfun$8$$anonfun$apply$1.apply(ClassToAPI.scala:78)
    at sbt.ClassToAPI$$anonfun$8$$anonfun$apply$1.apply(ClassToAPI.scala:78)
    at sbt.ClassToAPI$.fieldToDef(ClassToAPI.scala:159)
    at sbt.ClassToAPI$$anonfun$8.apply(ClassToAPI.scala:78)
    at sbt.ClassToAPI$$anonfun$8.apply(ClassToAPI.scala:78)
    at sbt.ClassToAPI$$anonfun$mergeMap$1.apply(ClassToAPI.scala:227)
    at sbt.ClassToAPI$$anonfun$mergeMap$1.apply(ClassToAPI.scala:227)
    at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:251)
    at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:251)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
    at scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:34)
    at scala.collection.TraversableLike$class.flatMap(TraversableLike.scala:251)
    at scala.collection.AbstractTraversable.flatMap(Traversable.scala:105)
    at sbt.ClassToAPI$.merge(ClassToAPI.scala:233)
    at sbt.ClassToAPI$.mergeMap(ClassToAPI.scala:227)
    at sbt.ClassToAPI$.structure(ClassToAPI.scala:78)
    at sbt.ClassToAPI$.x$2$lzycompute$1(ClassToAPI.scala:66)
    at sbt.ClassToAPI$.x$2$1(ClassToAPI.scala:66)
    at sbt.ClassToAPI$.instance$lzycompute$1(ClassToAPI.scala:66)
    at sbt.ClassToAPI$.sbt$ClassToAPI$$instance$1(ClassToAPI.scala:66)
    at sbt.ClassToAPI$$anonfun$5.apply(ClassToAPI.scala:67)
    at sbt.ClassToAPI$$anonfun$5.apply(ClassToAPI.scala:67)
    at xsbti.SafeLazy$Impl._t$lzycompute(SafeLazy.scala:18)
    at xsbti.SafeLazy$Impl._t(SafeLazy.scala:16)
    at xsbti.SafeLazy$Impl.get(SafeLazy.scala:22)
    at sbt.ClassToAPI$$anonfun$process$1.apply(ClassToAPI.scala:22)
    at sbt.ClassToAPI$$anonfun$process$1.apply(ClassToAPI.scala:22)
    at scala.collection.immutable.List.foreach(List.scala:318)
    at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:32)
    at scala.collection.mutable.ListBuffer.foreach(ListBuffer.scala:45)
    at sbt.ClassToAPI$.process(ClassToAPI.scala:22)
    at sbt.compiler.javac.AnalyzingJavaCompiler.sbt$compiler$javac$AnalyzingJavaCompiler$$readAPI$1(AnalyzingJavaCompiler.scala:76)
    at sbt.compiler.javac.AnalyzingJavaCompiler$$anonfun$compile$2$$anonfun$apply$mcV$sp$2$$anonfun$apply$5.apply(AnalyzingJavaCompiler.scala:84)
    at sbt.compiler.javac.AnalyzingJavaCompiler$$anonfun$compile$2$$anonfun$apply$mcV$sp$2$$anonfun$apply$5.apply(AnalyzingJavaCompiler.scala:84)
    at sbt.classfile.Analyze$$anonfun$apply$9.apply(Analyze.scala:44)
    at sbt.classfile.Analyze$$anonfun$apply$9.apply(Analyze.scala:43)
    at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:772)
    at scala.collection.mutable.HashMap$$anonfun$foreach$1.apply(HashMap.scala:98)
    at scala.collection.mutable.HashMap$$anonfun$foreach$1.apply(HashMap.scala:98)
    at scala.collection.mutable.HashTable$class.foreachEntry(HashTable.scala:226)
    at scala.collection.mutable.HashMap.foreachEntry(HashMap.scala:39)
    at scala.collection.mutable.HashMap.foreach(HashMap.scala:98)
    at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:771)
    at sbt.classfile.Analyze$.apply(Analyze.scala:43)
    at sbt.compiler.javac.AnalyzingJavaCompiler$$anonfun$compile$2$$anonfun$apply$mcV$sp$2.apply(AnalyzingJavaCompiler.scala:84)
    at sbt.compiler.javac.AnalyzingJavaCompiler$$anonfun$compile$2$$anonfun$apply$mcV$sp$2.apply(AnalyzingJavaCompiler.scala:82)
    at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:772)
    at scala.collection.immutable.List.foreach(List.scala:318)
    at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:771)
    at sbt.compiler.javac.AnalyzingJavaCompiler$$anonfun$compile$2.apply$mcV$sp(AnalyzingJavaCompiler.scala:82)
    at sbt.compiler.javac.AnalyzingJavaCompiler$$anonfun$compile$2.apply(AnalyzingJavaCompiler.scala:82)
    at sbt.compiler.javac.AnalyzingJavaCompiler$$anonfun$compile$2.apply(AnalyzingJavaCompiler.scala:82)
    at sbt.compiler.javac.AnalyzingJavaCompiler.timed(AnalyzingJavaCompiler.scala:93)
    at sbt.compiler.javac.AnalyzingJavaCompiler.compile(AnalyzingJavaCompiler.scala:81)
    at sbt.compiler.MixedAnalyzingCompiler$$anonfun$compileJava$1$1.apply$mcV$sp(MixedAnalyzingCompiler.scala:60)
    at sbt.compiler.MixedAnalyzingCompiler$$anonfun$compileJava$1$1.apply(MixedAnalyzingCompiler.scala:60)
    at sbt.compiler.MixedAnalyzingCompiler$$anonfun$compileJava$1$1.apply(MixedAnalyzingCompiler.scala:60)
    at sbt.compiler.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:74)
    at sbt.compiler.MixedAnalyzingCompiler.compileJava$1(MixedAnalyzingCompiler.scala:59)
    at sbt.compiler.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:64)
    at sbt.compiler.IC$$anonfun$compileInternal$1.apply(IncrementalCompiler.scala:160)
    at sbt.compiler.IC$$anonfun$compileInternal$1.apply(IncrementalCompiler.scala:160)
    at sbt.inc.IncrementalCompile$$anonfun$doCompile$1.apply(Compile.scala:66)
    at sbt.inc.IncrementalCompile$$anonfun$doCompile$1.apply(Compile.scala:64)
    at sbt.inc.IncrementalCommon.cycle(IncrementalCommon.scala:31)
    at sbt.inc.Incremental$$anonfun$1.apply(Incremental.scala:62)
    at sbt.inc.Incremental$$anonfun$1.apply(Incremental.scala:61)
    at sbt.inc.Incremental$.manageClassfiles(Incremental.scala:89)
    at sbt.inc.Incremental$.compile(Incremental.scala:61)
    at sbt.inc.IncrementalCompile$.apply(Compile.scala:54)
    at sbt.compiler.IC$.compileInternal(IncrementalCompiler.scala:160)
    at sbt.compiler.IC$.incrementalCompile(IncrementalCompiler.scala:138)
    at sbt.Compiler$.compile(Compiler.scala:152)
    at sbt.Compiler$.compile(Compiler.scala:138)
    at sbt.Defaults$.sbt$Defaults$$compileIncrementalTaskImpl(Defaults.scala:853)
    at sbt.Defaults$$anonfun$compileIncrementalTask$1.apply(Defaults.scala:844)
    at sbt.Defaults$$anonfun$compileIncrementalTask$1.apply(Defaults.scala:842)
    at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
    at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
    at sbt.std.Transform$$anon$4.work(System.scala:63)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
    at sbt.Execute.work(Execute.scala:235)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
    at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
[error] (akka-protobuf/compile:compileIncremental) java.lang.IllegalStateException: Failed to parse class java.util.AbstractMap$SimpleImmutableEntry: this may mean your classfiles are corrupted. Please clean and try again.
[error] Total time: 175 s, completed Nov 9, 2015 4:46:00 PM
[debug] > shell
```
# Expectation

Compilation succeeds.
# Note

sbt compiles successfully akka/akka if #2085 and #2214 are reverted.

**Comment by Duhemm** [comment]
Minimal reproduction:

``` java
package com.example;
import java.util.AbstractMap;
abstract class MyMap<K, V> extends AbstractMap<K, V> { }
```

**Comment by Duhemm** [comment]
`java.util.AbstractMap<K, V>` defines two static inner classes: `AbstractMap.SimpleEntry<K,V>` and `AbstractMap.SimpleImmutableEntry<K,V>`. (I think that the bug would still manifest itself if these inner classes were not static)

`MyMap<K, V>` inherits them, and `ClassToAPI.structure` will try to read the classfile of `AbstractMap.SimpleEntry`, which is located in a JAR. This will fail.

What do we want to do? Ignore classes for which we cannot find the classfile? Not try to analyse inherited inner classes? At first glance, this second solution should not introduce problems thanks to the current implementation of the incremental compiler, but I think that we don't want to make this kind of assumptions.

**Comment by stuhood** [comment]
Will take a look today... Sorry for the trouble!

**Comment by stuhood** [comment]
I believe I have code to do this already implemented. Feel free to assignee this issue to me, as I am working on it now.

**Comment by stuhood** [comment]
Opened https://github.com/sbt/sbt/pull/2268 ... I'm working to reproduce the issue in the Akka repo, but I believe that the pull resolves it.

**Comment by stuhood** [comment]
I've confirmed that Akka compiles at that sha with #2268 applied.

**Comment by eed3si9n** [comment]
@stuhood Merged #2268. Thanks for the update
